pkg/reconciler/autoscaling/hpa/keda_hpa.go

```diff
@@ -0,0 +1,160 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hpa
+
+import (
+	"context"
+	"fmt"
+
+	v2 "k8s.io/api/autoscaling/v2"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	autoscalingv2listers "k8s.io/client-go/listers/autoscaling/v2"
+
+	nv1alpha1 "knative.dev/networking/pkg/apis/networking/v1alpha1"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/ptr"
+	pkgreconciler "knative.dev/pkg/reconciler"
+	autoscalingv1alpha1 "knative.dev/serving/pkg/apis/autoscaling/v1alpha1"
+	"knative.dev/serving/pkg/autoscaler/config/autoscalerconfig"
+	pareconciler "knative.dev/serving/pkg/client/injection/reconciler/autoscaling/v1alpha1/podautoscaler"
+	areconciler "knative.dev/serving/pkg/reconciler/autoscaling"
+	"knative.dev/serving/pkg/reconciler/autoscaling/config"
+
+	"knative.dev/autoscaler-keda/pkg/client/clientset/versioned"
+	kedav1alpha1 "knative.dev/autoscaler-keda/pkg/client/listers/keda/v1alpha1"
+	"knative.dev/autoscaler-keda/pkg/reconciler/autoscaling/hpa/resources"
+)
+
+// Reconciler implements the control loop for the HPA resources.
+type Reconciler struct {
+	*areconciler.Base
+
+	kubeClient kubernetes.Interface
+	kedaClient versioned.Interface
+	kedaLister kedav1alpha1.ScaledObjectLister
+	hpaLister  autoscalingv2listers.HorizontalPodAutoscalerLister
+}
+
+// Check that our Reconciler implements pareconciler.Interface
+var _ pareconciler.Interface = (*Reconciler)(nil)
+
+// ReconcileKind implements Interface.ReconcileKind.
+func (c *Reconciler) ReconcileKind(ctx context.Context, pa *autoscalingv1alpha1.PodAutoscaler) pkgreconciler.Event {
+	ctx, cancel := context.WithTimeout(ctx, pkgreconciler.DefaultTimeout)
+	defer cancel()
+
+	logger := logging.FromContext(ctx)
+	logger.Debug("PA exists")
+
+	var hpa *v2.HorizontalPodAutoscaler
+
+	dScaledObject := resources.MakeScaledObject(pa, config.FromContext(ctx).Autoscaler)
+	scaledObj, err := c.kedaLister.ScaledObjects(pa.Namespace).Get(dScaledObject.Name)
+	if errors.IsNotFound(err) {
+		logger.Infof("Creating Scaled Object %q", dScaledObject.Name)
+		if scaledObj, err = c.kedaClient.KedaV1alpha1().ScaledObjects(dScaledObject.Namespace).Create(ctx, dScaledObject, metav1.CreateOptions{}); err != nil {
+			pa.Status.MarkResourceFailedCreation("ScaledObject", dScaledObject.Name)
+			return fmt.Errorf("failed to create ScaledObject: %w", err)
+		}
+	} else if err != nil {
+		return fmt.Errorf("failed to get ScaledObject: %w", err)
+	} else if !metav1.IsControlledBy(scaledObj, pa) {
+		// Surface an error in the PodAutoscaler's status, and return an error.
+		pa.Status.MarkResourceNotOwned("ScaledObject", dScaledObject.Name)
+		return fmt.Errorf("PodAutoscaler: %q does not own ScaledObject: %q", pa.Name, dScaledObject.Name)
+	}
+	if !equality.Semantic.DeepEqual(dScaledObject.Spec, scaledObj.Spec) {
```

Comment:
Do we care about labels/annotations or not?

Reply:
I will check.

Reply:
Let's consider this as part of a follow up PR about managing scaledObjects (managed and non managed mode), status propagation etc.